### ISSUE TITLE
Focal loss calculation is wrong

### ISSUE THREAD

**Issue by Garfield-hr** [luca-medeiros/lightning-sam#52]
Hi, I tried out your project and it solved my problem perfectly.  
I noticed that in my dataset, the focal loss is too small to be displayed properly in the terminal.   
![image](https://github.com/luca-medeiros/lightning-sam/assets/56104495/4a0e77a9-e6bc-4f85-a780-e20bfe2bb418)
After some investigation, it appears that there might be an issue with the focal loss calculation.   
Here's a snippet of the current implementation:
```py
    def forward(self, inputs, targets, alpha=ALPHA, gamma=GAMMA, smooth=1):
        inputs = F.sigmoid(inputs)
        inputs = torch.clamp(inputs, min=0, max=1)
        #flatten label and prediction tensors
        inputs = inputs.view(-1)
        targets = targets.view(-1)

        BCE = F.binary_cross_entropy(inputs, targets, reduction='mean')
        BCE_EXP = torch.exp(-BCE)
        focal_loss = alpha * (1 - BCE_EXP)**gamma * BCE

        return focal_loss
```
In the current implementation, a mean reduction is applied in the binary cross-entropy (BCE) calculation, resulting in BCE for the entire mask.  This leads to focus coefficients being calculated at the instance level. To better focus on difficult pixels and calculate the focal loss correctly, the focal loss calculation should be updated as follows:
``` py
    def forward(self, inputs, targets, alpha=ALPHA, gamma=GAMMA, smooth=1):
        inputs = F.sigmoid(inputs)
        inputs = torch.clamp(inputs, min=0, max=1)
        #flatten label and prediction tensors
        inputs = inputs.view(-1)
        targets = targets.view(-1)

        BCE = F.binary_cross_entropy(inputs, targets, reduction='none')
        BCE_EXP = torch.exp(-BCE)
        focal_loss = alpha * (1 - BCE_EXP)**gamma * BCE
        focal_loss = focal_loss.mean()

        return focal_loss
```
With this revision, the focal loss now focuses on difficult pixels as intended. This approach aligns with the focal loss implementation in the [MONAI library](https://docs.monai.io/en/1.3.0/losses.html#focalloss).

After applying this correction, the image below demonstrates the losses during the training process on my dataset.
![image](https://github.com/luca-medeiros/lightning-sam/assets/56104495/3c4646e9-57c6-4be6-8e03-993827ea02e3)

Feel free to incorporate these changes https://github.com/luca-medeiros/lightning-sam/pull/53 !
